### ISSUE TITLE
Jupyter Notebooks: npm import - `typeof document` expression fails with `document is not defined` error

### ISSUE THREAD

**Issue by nesterow** [denoland/deno#25872]
Version: Deno 1.46.3
Version: Deno 2.0.0-rc.5

In my Jupyter notebook I am using `vega` as follows:

```typescript
import * as p from './plots.ts'
```
I am getting following error on module load:
```bash
ReferenceError: document is not defined
    at domCanvas (file:///Users/user/Library/Caches/deno/npm/registry.npmjs.org/vega-canvas/1.2.7/build/vega-canvas.node.js:8:5)
    at Object.canvas (file:///Users/user/Library/Caches/deno/npm/registry.npmjs.org/vega-canvas/1.2.7/build/vega-canvas.node.js:42:12)
    at file:///Users/user/Library/Caches/deno/npm/registry.npmjs.org/vega-scenegraph/4.13.0/build/vega-scenegraph.js:2890:39
    at file:///Users/user/Library/Caches/deno/npm/registry.npmjs.org/vega-scenegraph/4.13.0/build/vega-scenegraph.js:2:66
    at Object.<anonymous> (file:///Users/user/Library/Caches/deno/npm/registry.npmjs.org/vega-scenegraph/4.13.0/build/vega-scenegraph.js:5:3)
    at Object.<anonymous> (file:///Users/user/Library/Caches/deno/npm/registry.npmjs.org/vega-scenegraph/4.13.0/build/vega-scenegraph.js:6524:4)
    at Module._compile (node:module:748:34)
    at Object.Module._extensions..js (node:module:767:10)
    at Module.load (node:module:665:32)
    at Function.Module._load (node:module:537:12)"
```

The following line fails on module load:
```typescript
// vega-canvas.node.js:8:5
if (typeof document !== 'undefined' && document.createElement)
```
And when I replace it with dummy false value, everything works as expected:
```typescript
// vega-canvas.node.js:8:5
if (false && document.createElement) 
```

`typeof document` expression fails, so I suspect the issue might be related to the parser. I was only able to confirm it on Jupyter kernel because currently I don't have other work in Deno, however other environments could be affected with the same module loader.

Importing libraries directly in the cell seems to work every time (unconfrmed):
```typescript
import * as vega from "npm:vega";
import * as lite from "npm:vega-lite";
```

The issue apears randomly after restarting the jupyter server and sometimes after restarting the kernel. Also, it may work in one notebook but fail in another. I couldn't find a reliable way to reproduce it - I update when I find it.

I've checked MacOS builds v1.46.3 and 2.0.0-rc.5 with Jupyter kernel using VSCode. 
Also confirmend on Linux builds with Jupyter in docker environment.

*UPDATE:*

The cell variables conflict with the global scope:
```typescript
const document = null;
import * as vega from "npm:vega";
```
```
Stack trace:
TypeError: Cannot read properties of null (reading 'createElement')
```
[Example](https://github.com/nesterow/shortcuts/blob/main/deno_jupyter_bug.ipynb)

**Comment by lucacasonato** [comment]
Can you log out `console.log(document)`? I can not reproduce this unfortunately

**Comment by nesterow** [comment]
I reproduced it on Linux. Importing from the local file:

```typescript
import * as p from './plots.ts'
```
<img width="1276" alt="Screenshot 2024-09-26 at 16 38 39" src="https://github.com/user-attachments/assets/8fd29a19-ae86-4f10-aabe-1226384c1f0c">

Inporting libraries directly in the cell seems to work every time:
```typescript
import * as vega from "npm:vega";
import * as lite from "npm:vega-lite";
``` 

The issue apears randomly after restarting the jupyter server and sometimes after restarting the kernel. Also, it may work in one notebook but fail in another. I couldn't find a reliable way to reproduce it - I update when I find it.

> Can you log out console.log(document)?

No, I cannot log it because it fails on the `typeof document` expression which must result to "undefined". The global document object must not be present on the server runtime.

**Comment by UrielCh** [comment]
same issue,

with a brand new deno kernel a simple:
```ts
console.log(document);
```
cause a `ReferenceError: document is not defined`

to try to make it works properly I removed my previous deno kernel an installed it using `deno 2.0.0 (stable, release, aarch64-apple-darwin)`

```bash
jupyter kernelspec uninstall deno
jupyter kernelspec list
deno jupyter --install
```
my new deno's `kernel.json` is:
```json
{
  "argv": [
    "/Users/uriel/.deno/bin/deno",
    "jupyter",
    "--kernel",
    "--conn",
    "{connection_file}"
  ],
  "display_name": "Deno",
  "language": "typescript"
}
```

**Comment by UrielCh** [comment]
note within the kernel `console.log(Object.keys(globalThis))` display:
```ts
[
  "Deno",                 "queueMicrotask",
  "atob",                 "btoa",
  "createImageBitmap",    "caches",
  "crypto",               "fetch",
  "EventSource",          "process",
  "reportError",          "structuredClone",
  "onunhandledrejection", "navigator",
  "alert",                "confirm",
  "prompt",               "localStorage",
  "sessionStorage",       "name",
  "close",                "closed",
  "onerror",              "onload",
  "onbeforeunload",       "onunload",
  "clear",                "clearInterval",
  "clearTimeout",         "performance",
  "setInterval",          "setTimeout",
  "self"
]
```
`windows` and `document` are not defined.

**Comment by nesterow** [comment]
The issue is related to the way deno repl works. I don't think deno repl should handle such cases. Perhaps I have to close the issue.
Just import vega or other modules in separate cell and everything should work.

**Comment by bartlomieju** [comment]
@nesterow sorry for a slow response, I'm still interested in getting to the bottom of this. `document` should always result in `ReferenceError` in Deno, unless a library provides it in some way.

**Comment by nesterow** [comment]
@bartlomieju I don't think it needs to be fixed.

As far as I understand Jupyter kernel uses deno repl under the hood, so all variables defined in the notebook cells are in the global scope. Try following in the deno repl: 
```javascript
let document = null
import "npm:vega"
```
The document will be in the global scope. Then try without defining document variable. 
Deno repl handles basic cases like not allowing to rewrite important globals, but I don't think it should do anything beyond it. It could be a task for repl improvement though.

The issue was that I imported `document` in the same time I imported the library that checks if there is a document in the global scope. So the document variable got referenced in the global scope, but hasn't yet been initialized because the modules were resolving. The issue appeared random to me because it is easy to forget in what order you run the cells in the notebook when you constanly fix something.

Because all variables defined in Jupyter notebook cells are global, the solution is to import all dependecies before doing anything (as they do it in python). 

I mostly used it to port some old python code to javascript, so currently deno's kernel covers everything I need. I have [some test notebooks](https://github.com/nesterow/shortcuts/tree/main/notebooks) - it might help if anyone needs to do similar task.
